File: part-02/src/main/java/eu/luminis/workshop/smallsteps/logic/appservice/LegoStockQueries.java
package eu.luminis.workshop.smallsteps.logic.appservice;

import eu.luminis.workshop.smallsteps.logic.domainmodel.valueobjects.LegoStoreId;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class LegoStockQueries {
    private final LegoStockRepository repository;

    @Autowired
    public LegoStockQueries(LegoStockRepository repository) {
        this.repository = repository;
    }

    public Map<String, Integer> currentlyMissingPartsReport(LegoStoreId legoStoreId) {
        StockState foundState = repository.find(legoStoreId);

        List<Map<String, Integer>> collect = foundState.getIncompleteStock().stream()
                .map(LegoBox::getMissingParts)
                .collect(Collectors.toList());

        Map<String, Integer> merged = mergeAndSum(collect);

        return toSortedMap(merged);
    }

    public Map<String, Integer> historicallyMostLostParts(LegoStoreId legoStoreId) {
        StockState foundState = repository.find(legoStoreId);

        List<Map<String, Integer>> collect = foundState.getIncompleteReturnHistory().stream()
                .map(IncompleteReturn::getMissingParts)
                .collect(Collectors.toList());

        Map<String, Integer> merged = mergeAndSum(collect);

        return toSortedMap(merged);

    }

    private Map<String, Integer> mergeAndSum(List<Map<String, Integer>> collect) {
        Map<String, Integer> missingParts = new HashMap<>();
        collect.forEach(item -> {
            item.keySet().forEach(key -> {
                missingParts.merge(key, item.get(key), Integer::sum);
            });
        });
        return missingParts;
    }

    private Map<String, Integer> toSortedMap(Map<String, Integer> missingParts) {
        return missingParts.entrySet().stream()
                .sorted(Comparator.comparing(o -> Integer.valueOf(o.getKey())))
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        Map.Entry::getValue,
                        (v1, v2) -> {
                            throw new RuntimeException(String.format("Duplicate key for values %s and %s", v1, v2));
                        },
                        LinkedHashMap::new));
    }
}

File: part-02/src/main/kotlin/eu/luminis/workshop/smallsteps/logic/appService/LegoStockQueries.kt
package eu.luminis.workshop.smallsteps.logic.appService

import eu.luminis.workshop.smallsteps.logic.domainModel.LegoParts
import eu.luminis.workshop.smallsteps.logic.domainModel.toLegoParts
import eu.luminis.workshop.smallsteps.logic.domainModel.valueObjects.LegoStoreId

class LegoStockQueries(private val repository: LegoStockRepository) {
    fun currentlyMissingPartsReport(store: LegoStoreId): List<Pair<String, Int>> {
        val stockState = repository.find(store)
        return stockState.incompleteStock
            .map { it.missingParts.toLegoParts() }
            .mergeAndSum()
            .toSortedListOfPairs()
    }

    fun historicallyMostLostParts(store: LegoStoreId): List<Pair<String, Int>> {
        val stockState = repository.find(store)
        return stockState.incompleteReturnHistory
            .map { it.missingParts.toLegoParts() }
            .mergeAndSum()
            .toSortedListOfPairs()
    }
}

private fun List<LegoParts>.mergeAndSum(): LegoParts {
    return this.fold(LegoParts(), LegoParts::plus)
}

private fun LegoParts.toSortedListOfPairs(): List<Pair<String, Int>> {
    return this.toListOfPartCounts().sortedBy { it.first }.sortedByDescending { it.second }
}
